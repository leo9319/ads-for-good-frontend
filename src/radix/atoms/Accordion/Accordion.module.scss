@import '../../../styles/breakpoints.scss';
@import '../../../styles/colours.scss';

.accordionContainer {
  background-color: var(--backgroundColor);
  &.accordionPadding {
    padding: var(--spacing-lg);
    box-sizing: border-box;
  }
  &.accordionPaddingRounded {
    border-radius: 16px;
  }
  .accordionTitleWrapper {
    margin: 0;
    .accordionTitle {
      width: 100%;
      padding: 0px;
      box-sizing: border-box;
      display: flex;
      align-items: center;
      justify-content: space-between;
      background: transparent;
      border: none;
      cursor: pointer;
      transition: all 0.3s ease;
      &[data-state='closed'] .accordionToggleIcon {
        transform: rotate(180deg);
        transform-origin: center;
        transition: all 0.3s ease;
      }
      &[data-state='open'] {
        margin-bottom: var(--spacing-md);
        transition: all 0.3s ease;
      }
      .accordionTitleName {
        font-size: var(--typography-font-size-4);
        line-height: var(--typography-font-size-6);
        font-weight: 600;
        color: $color-neutral-950;
        text-align: left;
      }
      .accordionToggleIconScetion {
        .accordionToggleIcon {
          transition: all 0.3s ease;
        }
      }
    }
  }
  .accordionContentWrapper {
    width: calc(100% - 32px); // This 32px is toggle icon size
    overflow: hidden;
    transition: all 0.3s ease;
    &[data-state='open'] {
      animation: accordion-down 0.3s ease;
    }
    &[data-state='closed'] {
      animation: accordion-up 0.3s ease;
    }
    .accordionContent {
      font-size: var(--typography-font-size-3);
      line-height: var(--typography-font-size-5);
      font-weight: 400;
      color: $color-neutral-600;
    }
  }
  .divider {
    width: 100% !important;
    margin: var(--spacing-md) 0 !important;
    border: 1px solid $color-neutral-300;
  }
}

@keyframes accordion-down {
  from {
    height: 0;
    opacity: 0;
  }
  to {
    height: var(--radix-accordion-content-height);
    opacity: 1;
  }
}

@keyframes accordion-up {
  from {
    height: var(--radix-accordion-content-height);
    opacity: 1;
  }
  to {
    height: 0;
    opacity: 0;
  }
}

@media (max-width: $md-min) {
  .accordionContainer {
    &.accordionPadding {
      padding: var(--spacing-md);
    }
    .accordionTitleWrapper {
      .accordionTitle {
        .accordionTitleName {
          font-size: var(--typography-font-size-2);
          line-height: var(--typography-font-size-3);
        }
      }
    }
    .accordionContentWrapper {
      .accordionContent {
        font-size: var(--typography-font-size-2);
        line-height: var(--typography-font-size-3);
      }
    }
  }
}
