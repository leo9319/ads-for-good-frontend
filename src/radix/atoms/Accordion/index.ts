export type { AccordionProps } from './Accordion';
export type { AccordionSingleItemProps } from './Accordion';
export { Accordion } from './Accordion';
export { default } from './Accordion';
