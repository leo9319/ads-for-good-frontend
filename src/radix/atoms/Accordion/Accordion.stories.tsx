import React from 'react';
import Accordion from './Accordion';
import { Meta, StoryObj } from '@storybook/react';
import { Theme } from '@radix-ui/themes';

const meta = {
  title: 'Library/Atoms/Accordion',
  component: Accordion,
  decorators: [
    Story => (
      <Theme style={{ padding: '10px', boxSizing: 'border-box' }}>
        <Story />
      </Theme>
    ),
  ],
  parameters: {
    layout: 'fullscreen',
  },
  tags: ['autodocs'],
  argTypes: {
    className: {
      control: 'text',
    },
    type: {
      control: 'select',
      options: ['single', 'multiple'],
    },
    collapsible: {
      control: 'boolean',
    },
    applyRounded: {
      control: 'boolean',
    },
    applyPadding: {
      control: 'boolean',
    },
    backgroundColor: {
      control: 'color',
    },
  },
} satisfies Meta<typeof Accordion>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Accordions: Story = {
  args: {
    className: '',
    type: 'multiple',
    collapsible: true,
    applyRounded: false,
    applyPadding: false,
    backgroundColor: '',
    items: [
      {
        value: 'item-1',
        title: 'Getting Started with the Basics',
        content:
          'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque non risus in orci varius accumsan. Integer nec odio. Praesent libero. Sed cursus ante dapibus diam.',
      },
      {
        value: 'item-2',
        title: 'Advanced Techniques and Tips',
        content:
          'Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit.',
      },
      {
        value: 'item-3',
        title: 'Troubleshooting Common Issues',
        content:
          'Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum iure reprehenderit.',
      },
      {
        value: 'item-4',
        title: 'Best Practices and Final Thoughts',
        content:
          'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident.',
      },
    ],
  },
};
