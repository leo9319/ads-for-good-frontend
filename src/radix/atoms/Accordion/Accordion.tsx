import React, { ComponentPropsWithoutRef, ReactNode } from 'react';
import { Accordion as RadixAccordion } from 'radix-ui';
import { classNames, getModuleClasses } from '@utils/common/classNames';
import { Icon } from '@components/atoms/Icons';
import { getStylesAsCssProperties } from '@utils/common/styles';
import { truncateByCharacterCount } from '@utils/common/truncateByCharacterCount';

import styles from './Accordion.module.scss';

/**
 * @interface
 * The Accordion props type of {@link Accordion}
 */
export type AccordionSingleItemProps = {
  /**
   * value for the Accordion
   */
  value: string;
  /**
   * title for the Accordion
   */
  title: React.ReactNode;
  /**
   * content for the Accordion
   */
  content: React.ReactNode;
};

export type AccordionProps = {
  /**
   * class name for the Accordion wrapper
   */
  className?: string;
  /**
   * Items to render in the accordion
   */
  items: AccordionSingleItemProps[];
  /**
   * How many item to be opened at once
   */
  type?: 'single' | 'multiple';
  /**
   * To define atleast one accordion to be open always, this wont work on multiple
   */
  collapsible?: boolean;
  /**
   * Default value for expanded items
   */
  defaultValue?: string | string[];
  /**
   * To set border radius to the accordion
   */
  applyRounded?: boolean;
  /**
   * To set padding to the accordion
   */
  applyPadding?: boolean;
  /**
   * To set background color for the accordion
   */
  backgroundColor?: string;
};

type CommonProps = {
  /**
   * Children to render in the accordion content
   */
  children: ReactNode;
  /**
   * class name for the Accordion content wrapper
   */
  className?: string;
};
type TriggerProps = ComponentPropsWithoutRef<typeof RadixAccordion.Trigger> &
  CommonProps;
type ContentProps = ComponentPropsWithoutRef<typeof RadixAccordion.Content> &
  CommonProps;

/**
 *
 * Accordion component.
 *
 * @group Atoms
 * @category Radix Component
 *
 * @param {AccordionProps} props - The Accordion props
 * @returns {ReactElement} - Accordion Component
 *
 */
const RadixAccordionTrigger = React.forwardRef<HTMLButtonElement, TriggerProps>(
  ({ children, className, ...props }, forwardedRef) => (
    <RadixAccordion.Header
      className={classNames(
        styles.accordionTitleWrapper,
        'accordionTitleWrapper',
        className
      )}
    >
      <RadixAccordion.Trigger
        className={classNames(styles.accordionTitle, 'accordionTitle')}
        {...props}
        ref={forwardedRef}
      >
        <span
          className={classNames(
            styles.accordionTitleName,
            'accordionTitleName'
          )}
        >
          {children}
        </span>
        <span
          className={classNames(
            styles.accordionToggleIconScetion,
            'accordionToggleIconScetion'
          )}
        >
          <Icon
            name="icon-chevron"
            className={classNames(
              styles.accordionToggleIcon,
              'accordionToggleIcon'
            )}
            backgroundSize="32px"
            size="16"
            color="#17191F"
            backgroundColor="#fff"
            rounded={true}
          ></Icon>
        </span>
      </RadixAccordion.Trigger>
    </RadixAccordion.Header>
  )
);
RadixAccordionTrigger.displayName = 'RadixAccordionTrigger';

const RadixAccordionContent = React.forwardRef<HTMLDivElement, ContentProps>(
  ({ children, className, ...props }, forwardedRef) => {
    return (
      <RadixAccordion.Content
        className={classNames(
          styles.accordionContentWrapper,
          'accordionContentWrapper',
          className
        )}
        {...props}
        ref={forwardedRef}
      >
        <div
          className={classNames(styles.accordionContent, 'accordionContent')}
        >
          {children}
        </div>
      </RadixAccordion.Content>
    );
  }
);
RadixAccordionContent.displayName = 'RadixAccordionContent';

type AccordionItemProps = {
  item: AccordionSingleItemProps;
  index: number;
  totalItems: number;
};
const AccordionItem = ({ item, index, totalItems }: AccordionItemProps) => (
  <>
    <RadixAccordion.Item key={item.value} value={item.value}>
      <RadixAccordionTrigger>
        {truncateByCharacterCount(item.title, 150)}
      </RadixAccordionTrigger>
      <RadixAccordionContent>
        {truncateByCharacterCount(item.content, 500)}
      </RadixAccordionContent>
    </RadixAccordion.Item>
    {index < totalItems && (
      <div className={classNames(styles.divider, 'divider')} />
    )}
  </>
);
AccordionItem.displayName = 'AccordionItem';

type ItemsProps = {
  items: AccordionSingleItemProps[];
  totalItems: number;
};
const Items = ({ items, totalItems }: ItemsProps) =>
  items.map((item, index) => (
    <AccordionItem
      key={item.value}
      item={item}
      index={index}
      totalItems={totalItems}
    />
  ));
AccordionItem.displayName = 'Items';

export const Accordion = ({
  className,
  items,
  collapsible = true,
  type = 'multiple',
  defaultValue,
  applyPadding,
  applyRounded,
  backgroundColor = '',
}: AccordionProps) => {
  const classes = getModuleClasses(className?.trim(), styles);
  const wrapperClass = classNames(
    styles.accordionContainer,
    'accordionContainer',
    applyPadding ? styles.accordionPadding : '',
    applyRounded ? styles.accordionPaddingRounded : '',
    classes
  );

  if (items?.length <= 0) {
    return null;
  }

  const firstValue = items[0]?.value;
  const defaultStringValue = (
    type === 'single' ? (defaultValue ?? firstValue) : ''
  ) as string;
  const defaultStringArrayValue = (
    type === 'multiple' ? (defaultValue ?? [firstValue]) : []
  ) as string[];
  const totalItems = items.length - 1;

  return (
    <div
      className={wrapperClass}
      style={getStylesAsCssProperties({ '--backgroundColor': backgroundColor })}
    >
      {type === 'single' ? (
        <RadixAccordion.Root
          type="single"
          collapsible={collapsible}
          defaultValue={defaultStringValue}
        >
          <Items items={items} totalItems={totalItems} />
        </RadixAccordion.Root>
      ) : (
        <RadixAccordion.Root
          type="multiple"
          defaultValue={defaultStringArrayValue}
        >
          <Items items={items} totalItems={totalItems} />
        </RadixAccordion.Root>
      )}
    </div>
  );
};

export default Accordion;
