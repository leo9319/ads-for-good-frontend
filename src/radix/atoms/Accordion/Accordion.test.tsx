import React from 'react';
import Accordion from './Accordion';
import { render } from '@testing-library/react';

describe('Accordion', () => {
  it('should render successfully', () => {
    const { baseElement } = render(
      <Accordion
        items={[
          {
            value: 'item-1',
            title: 'Getting Started with the Basics',
            content:
              'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque non risus in orci varius accumsan. Integer nec odio. Praesent libero. Sed cursus ante dapibus diam.',
          },
          {
            value: 'item-2',
            title: 'Advanced Techniques and Tips',
            content:
              'Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit.',
          },
          {
            value: 'item-3',
            title: 'Troubleshooting Common Issues',
            content:
              'Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum iure reprehenderit.',
          },
          {
            value: 'item-4',
            title: 'Best Practices and Final Thoughts',
            content:
              'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident.',
          },
        ]}
      />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should not render successfully when the items are not available', () => {
    const { baseElement } = render(<Accordion items={[]} />);
    expect(baseElement).toMatchSnapshot();
  });
});
