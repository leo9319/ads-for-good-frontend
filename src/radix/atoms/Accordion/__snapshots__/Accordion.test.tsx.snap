// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should not render successfully when the items are not available 1`] = `
<body>
  <div />
</body>
`;

exports[`Accordion should render successfully 1`] = `
<body>
  <div>
    <div
      class="accordionContainer accordionContainer"
    >
      <div
        data-orientation="vertical"
      >
        <div
          data-orientation="vertical"
          data-state="open"
        >
          <h3
            class="accordionTitleWrapper accordionTitleWrapper"
            data-orientation="vertical"
            data-state="open"
          >
            <button
              aria-controls="radix-:r1:"
              aria-expanded="true"
              class="accordionTitle accordionTitle"
              data-orientation="vertical"
              data-radix-collection-item=""
              data-state="open"
              id="radix-:r0:"
              type="button"
            >
              <span
                class="accordionTitleName accordionTitleName"
              >
                Getting Started with the Basics
              </span>
              <span
                class="accordionToggleIconScetion accordionToggleIconScetion"
              >
                <span
                  class="accordionToggleIcon accordionToggleIcon"
                  style="background: rgb(255, 255, 255); border-radius: 50%; width: 32px; height: 32px; display: grid; align-items: center; justify-content: center;"
                >
                  <span
                    class="icon icon-chevron"
                    style="width: 16px; height: 16px; color: rgb(23, 25, 31);"
                  />
                </span>
              </span>
            </button>
          </h3>
          <div
            aria-labelledby="radix-:r0:"
            class="accordionContentWrapper accordionContentWrapper"
            data-orientation="vertical"
            data-state="open"
            id="radix-:r1:"
            role="region"
            style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width); transition-duration: 0s; animation-name: none;"
          >
            <div
              class="accordionContent accordionContent"
            >
              Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque non risus in orci varius accumsan. Integer nec odio. Praesent libero. Sed cursus ante dapibus diam.
            </div>
          </div>
        </div>
        <div
          class="divider divider"
        />
        <div
          data-orientation="vertical"
          data-state="closed"
        >
          <h3
            class="accordionTitleWrapper accordionTitleWrapper"
            data-orientation="vertical"
            data-state="closed"
          >
            <button
              aria-controls="radix-:r3:"
              aria-expanded="false"
              class="accordionTitle accordionTitle"
              data-orientation="vertical"
              data-radix-collection-item=""
              data-state="closed"
              id="radix-:r2:"
              type="button"
            >
              <span
                class="accordionTitleName accordionTitleName"
              >
                Advanced Techniques and Tips
              </span>
              <span
                class="accordionToggleIconScetion accordionToggleIconScetion"
              >
                <span
                  class="accordionToggleIcon accordionToggleIcon"
                  style="background: rgb(255, 255, 255); border-radius: 50%; width: 32px; height: 32px; display: grid; align-items: center; justify-content: center;"
                >
                  <span
                    class="icon icon-chevron"
                    style="width: 16px; height: 16px; color: rgb(23, 25, 31);"
                  />
                </span>
              </span>
            </button>
          </h3>
          <div
            aria-labelledby="radix-:r2:"
            class="accordionContentWrapper accordionContentWrapper"
            data-orientation="vertical"
            data-state="closed"
            hidden=""
            id="radix-:r3:"
            role="region"
            style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
          />
        </div>
        <div
          class="divider divider"
        />
        <div
          data-orientation="vertical"
          data-state="closed"
        >
          <h3
            class="accordionTitleWrapper accordionTitleWrapper"
            data-orientation="vertical"
            data-state="closed"
          >
            <button
              aria-controls="radix-:r5:"
              aria-expanded="false"
              class="accordionTitle accordionTitle"
              data-orientation="vertical"
              data-radix-collection-item=""
              data-state="closed"
              id="radix-:r4:"
              type="button"
            >
              <span
                class="accordionTitleName accordionTitleName"
              >
                Troubleshooting Common Issues
              </span>
              <span
                class="accordionToggleIconScetion accordionToggleIconScetion"
              >
                <span
                  class="accordionToggleIcon accordionToggleIcon"
                  style="background: rgb(255, 255, 255); border-radius: 50%; width: 32px; height: 32px; display: grid; align-items: center; justify-content: center;"
                >
                  <span
                    class="icon icon-chevron"
                    style="width: 16px; height: 16px; color: rgb(23, 25, 31);"
                  />
                </span>
              </span>
            </button>
          </h3>
          <div
            aria-labelledby="radix-:r4:"
            class="accordionContentWrapper accordionContentWrapper"
            data-orientation="vertical"
            data-state="closed"
            hidden=""
            id="radix-:r5:"
            role="region"
            style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
          />
        </div>
        <div
          class="divider divider"
        />
        <div
          data-orientation="vertical"
          data-state="closed"
        >
          <h3
            class="accordionTitleWrapper accordionTitleWrapper"
            data-orientation="vertical"
            data-state="closed"
          >
            <button
              aria-controls="radix-:r7:"
              aria-expanded="false"
              class="accordionTitle accordionTitle"
              data-orientation="vertical"
              data-radix-collection-item=""
              data-state="closed"
              id="radix-:r6:"
              type="button"
            >
              <span
                class="accordionTitleName accordionTitleName"
              >
                Best Practices and Final Thoughts
              </span>
              <span
                class="accordionToggleIconScetion accordionToggleIconScetion"
              >
                <span
                  class="accordionToggleIcon accordionToggleIcon"
                  style="background: rgb(255, 255, 255); border-radius: 50%; width: 32px; height: 32px; display: grid; align-items: center; justify-content: center;"
                >
                  <span
                    class="icon icon-chevron"
                    style="width: 16px; height: 16px; color: rgb(23, 25, 31);"
                  />
                </span>
              </span>
            </button>
          </h3>
          <div
            aria-labelledby="radix-:r6:"
            class="accordionContentWrapper accordionContentWrapper"
            data-orientation="vertical"
            data-state="closed"
            hidden=""
            id="radix-:r7:"
            role="region"
            style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
